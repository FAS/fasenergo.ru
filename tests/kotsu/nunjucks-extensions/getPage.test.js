/* eslint-env jest */

import { renderString } from '../../utils/nunjucks'

const render = (template, context = mockContext, parse) => renderString(template, context, parse)
const mockContext = {
  SITE: {
    __matter: {
      'index': {
        'props': { 'url': '/' }
      },
      'blog': {
        'props': { 'url': '/blog' },
        '2015-10-12-example-article': {
          'props': { 'url': '/blog/2015-10-12-example-article' }
        }
      },
      'testVar': {
        'props': { 'var': '__globarvar value: {{ __globalvar }}' }
      },
      'testFunc': {
        'props': { 'func': 'Crumbled url: {{ crumble("blog") }}' }
      }
      // @todo Disabled, because it will always fail due to endless recursion in cached `getPage`
      // 'testGetPage': {
      //   'props': { 'func': 'Blog url: {{ getPage("blog").props.url }}' }
      // }
    }
  }
}

describe('Nunjucks global function `getPage()`', () => {
  describe('should get from', () => {
    it('root string-based path', () => {
      expect(render(`{{ getPage('blog')|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('root', () => {
      expect(render(`{{ getPage('/').props|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('root string-based path with leading slash', () => {
      expect(render(`{{ getPage('/blog')|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('root string-based path with trailing slash', () => {
      expect(render(`{{ getPage('blog/')|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('root string-based path with leading and trailing slashes', () => {
      expect(render(`{{ getPage('/blog/')|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('root Array-based path', () => {
      expect(render(`{{ getPage(['blog'])|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('nested dot-notation path', () => {
      expect(render(`{{ getPage('blog.2015-10-12-example-article').props|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('nested Array-based path', () => {
      expect(render(`{{ getPage(['blog', '2015-10-12-example-article']).props|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('nested url-like path', () => {
      expect(render(`{{ getPage('blog/2015-10-12-example-article').props|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('nested url-like path with leading slash', () => {
      expect(render(`{{ getPage('/blog/2015-10-12-example-article').props|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('nested url-like path with trailing slash', () => {
      expect(render(`{{ getPage('blog/2015-10-12-example-article/').props|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })

    it('nested url-like path with leading and trailing slash', () => {
      expect(render(`{{ getPage('/blog/2015-10-12-example-article/').props|dump|safe }}`, undefined, true)).toMatchSnapshot()
    })
  })

  describe('should force-render received data', () => {
    it('with current context', () => {
      expect(render(`{% set __globalvar = 'testing __globalvar value' %}{{ getPage('testVar').props.var }}`)).toMatchSnapshot()
    })

    it('with current custom functions', () => {
      expect(render(`{{ getPage('testFunc').props.func }}`)).toMatchSnapshot()
    })
  })
})
