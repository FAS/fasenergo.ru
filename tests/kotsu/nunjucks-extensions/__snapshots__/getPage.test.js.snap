// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nunjucks global function \`getPage()\` should force-render received data with current context 1`] = `"__globarvar value: testing __globalvar value"`;

exports[`Nunjucks global function \`getPage()\` should force-render received data with current custom functions 1`] = `"Crumbled url: blog"`;

exports[`Nunjucks global function \`getPage()\` should get from nested Array-based path 1`] = `
Object {
  "url": "/blog/2015-10-12-example-article",
}
`;

exports[`Nunjucks global function \`getPage()\` should get from nested dot-notation path 1`] = `
Object {
  "url": "/blog/2015-10-12-example-article",
}
`;

exports[`Nunjucks global function \`getPage()\` should get from nested url-like path 1`] = `
Object {
  "url": "/blog/2015-10-12-example-article",
}
`;

exports[`Nunjucks global function \`getPage()\` should get from nested url-like path with leading and trailing slash 1`] = `
Object {
  "url": "/blog/2015-10-12-example-article",
}
`;

exports[`Nunjucks global function \`getPage()\` should get from nested url-like path with leading slash 1`] = `
Object {
  "url": "/blog/2015-10-12-example-article",
}
`;

exports[`Nunjucks global function \`getPage()\` should get from nested url-like path with trailing slash 1`] = `
Object {
  "url": "/blog/2015-10-12-example-article",
}
`;

exports[`Nunjucks global function \`getPage()\` should get from root 1`] = `
Object {
  "url": "/",
}
`;

exports[`Nunjucks global function \`getPage()\` should get from root Array-based path 1`] = `
Object {
  "2015-10-12-example-article": Object {
    "props": Object {
      "url": "/blog/2015-10-12-example-article",
    },
  },
}
`;

exports[`Nunjucks global function \`getPage()\` should get from root string-based path 1`] = `
Object {
  "2015-10-12-example-article": Object {
    "props": Object {
      "url": "/blog/2015-10-12-example-article",
    },
  },
}
`;

exports[`Nunjucks global function \`getPage()\` should get from root string-based path with leading and trailing slashes 1`] = `
Object {
  "2015-10-12-example-article": Object {
    "props": Object {
      "url": "/blog/2015-10-12-example-article",
    },
  },
}
`;

exports[`Nunjucks global function \`getPage()\` should get from root string-based path with leading slash 1`] = `
Object {
  "2015-10-12-example-article": Object {
    "props": Object {
      "url": "/blog/2015-10-12-example-article",
    },
  },
}
`;

exports[`Nunjucks global function \`getPage()\` should get from root string-based path with trailing slash 1`] = `
Object {
  "2015-10-12-example-article": Object {
    "props": Object {
      "url": "/blog/2015-10-12-example-article",
    },
  },
}
`;
